hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/DemoteDataMemberTest.java

```diff
@@ -0,0 +1,516 @@
+    }
+
+    @Test
+    public void demotion_shouldFail_whenClusterStateFrozen() {
+        demotion_shouldFail_whenClusterState_NotAllowMigration(ClusterState.FROZEN);
+    }
+
+    @Test
+    public void demotion_shouldFail_whenClusterStateNoMigration() {
+        demotion_shouldFail_whenClusterState_NotAllowMigration(ClusterState.NO_MIGRATION);
+    }
+
+    private void demotion_shouldFail_whenClusterState_NotAllowMigration(ClusterState state) {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();
+
+        HazelcastInstance hz1 = factory.newHazelcastInstance(new Config());
+        HazelcastInstance hz2 = factory.newHazelcastInstance(new Config());
+        HazelcastInstance hz3 = factory.newHazelcastInstance(new Config());
+
+        warmUpPartitions(hz1, hz2, hz3);
+        changeClusterStateEventually(hz2, state);
+
+        exception.expect(IllegalStateException.class);
+        hz2.getCluster().demoteLocalDataMember();
+    }
+
+    @Test
+    public void demotion_shouldFail_whenMastershipClaimInProgress_duringDemotion() {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();
+
+        HazelcastInstance hz1 = factory.newHazelcastInstance(new Config());
+        HazelcastInstance hz2 = factory.newHazelcastInstance(new Config());
+
+        // artificially set mastership claim flag
+        ClusterServiceImpl clusterService = getNode(hz1).getClusterService();
+        clusterService.getClusterJoinManager().setMastershipClaimInProgress();
+
+        Cluster cluster = hz2.getCluster();
+        exception.expect(IllegalStateException.class);
+        cluster.demoteLocalDataMember();
+    }
+
+    @Test
+    public void demotion_shouldFail_whenMasterLeaves_duringDemotion() throws Exception {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();
+
+        HazelcastInstance hz1 = factory.newHazelcastInstance(new Config());
+        HazelcastInstance hz2 = factory.newHazelcastInstance(new Config());
+        HazelcastInstance hz3 = factory.newHazelcastInstance(new Config());
+
+        assertClusterSizeEventually(3, hz2);
+
+        dropOperationsBetween(hz3, hz1, F_ID, singletonList(DEMOTE_DATA_MEMBER));
+        Cluster cluster = hz3.getCluster();
+        Future<Exception> future = spawn(() -> {
+            try {
+                cluster.demoteLocalDataMember();
+            } catch (Exception e) {
+                return e;
+            }
+            return null;
+        });
+        assertDemotionInvocationStarted(hz3);
+
+        hz1.getLifecycleService().terminate();
+        assertClusterSizeEventually(2, hz2, hz3);
+
+        Exception exception = future.get();
+        // MemberLeftException is wrapped by HazelcastException
+        assertInstanceOf(MemberLeftException.class, exception.getCause());
+    }
+
+    @Test
+    public void demotion_shouldFail_whenMasterIsSuspected_duringDemotion() throws Exception {
```

Comment:
After I changed to using `#invokeOnMaster` in `ClusterServiceImpl#demoteLocalDataMember` I lost track of where the returned `MemberView` came from. We need to know that in order to make the validations in `ClusterServiceImpl#updateMembers`.

In https://github.com/hazelcast/hazelcast/pull/24617/commits/42a2a47db43c0dd9f6c1d0c993615a5412ab1d24 I solved this by changing response type of DemoteDataMemberOp to the new type `MembersViewResponse` which also includes the info about which member the `MemberView` comes from along with `MemberView` itself.

This _works_ but it makes the code a bit cluttered - e.g. promotion and demotion are slightly different in this aspect for no apparent reason.

Other possible solutions might be;
* Add such info inside `MemberView`. 
* Also change response type of `PromoteLiteMemberOp` to  `MembersViewResponse` and use `invokeOnMaster` (it would probably be challenging from a compatibility perspective)
* Just skip the validation in `ClusterServiceImpl#updateMembers` after we used invokeOnMaster?

What is your thoughts on this?